### PR TITLE
Replace `thrust::tuple` implementation with `cuda::std::tuple`

### PR DESCRIPTION
**Pull request by miscco**
Our old implementation of `tuple` has a lot of limitations.

So rather than using the old homegrown version reuse `cuda::std::tuple`

### CODE REVIEW

File: thrust/thrust/tuple.h [hunk 1]
```diff
@@ -178,390 +108,12 @@ get(const detail::cons<HT, TT>& t);
...
- *
- *  \param x The first \p tuple to swap.
- *  \param y The second \p tuple to swap.
- */
-template<
-  typename T0, typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9,
-  typename U0, typename U1, typename U2, typename U3, typename U4, typename U5, typename U6, typename U7, typename U8, typename U9
->
-inline __host__ __device__
-void swap(tuple<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> &x,
```
Comment (senior-zero) [inline thrust/thrust/tuple.h:467]: The following code used to work, but now it doesn't:
```cpp
  thrust::swap(a, b);
```

File: thrust/thrust/pair.h [hunk 2]
```diff
@@ -33,242 +34,43 @@ THRUST_NAMESPACE_BEGIN
...
-
-
-/*! \p swap swaps the contents of two <tt>pair</tt>s.
- *
- *  \param x The first \p pair to swap.
- *  \param y The second \p pair to swap.
- */
-template <typename T1, typename T2>
-  inline __host__ __device__
-    void swap(pair<T1,T2> &x, pair<T1,T2> &y);
```
Comment (senior-zero) [inline thrust/thrust/pair.h:206]: The following code used to work, now it doesn't:

```cpp
  thrust::pair<int, float> a(1, 2.0f);
  thrust::pair<int, float> b(2, 1.0f);

  thrust::swap(a, b);
```

File: thrust/thrust/tuple.h [hunk 3]
```diff
@@ -71,73 +66,8 @@ template <size_t N, class T> struct tuple_element;
...
- *
- *  \see pair
- *  \see tuple
- */
-template<int N, class HT, class TT>
-__host__ __device__
-inline typename access_traits<
-                  typename tuple_element<N, detail::cons<HT, TT> >::type
-                >::non_const_type
-get(detail::cons<HT, TT>& t);
```
Comment (senior-zero) [inline thrust/thrust/tuple.h:107]: The following code used to work, now it doesn't:
```cpp
  thrust::pair<int, float> a(1, 2.0f);
  thrust::get<0>(a);
```

File: thrust/testing/tuple_transform.cu [hunk 4]
```diff
@@ -19,10 +19,7 @@ struct GetFunctor
 {
   template<typename Tuple>
   __host__ __device__
-  typename thrust::access_traits<
-    typename thrust::tuple_element<N, Tuple>::type
-  >::const_type
-  operator()(const Tuple &t)
+  decltype(auto) operator()(const Tuple &t)
```
Comment (senior-zero) [inline thrust/testing/tuple_transform.cu:22]: Isn't `decltype(auto)` a C++14 feature?

**Review by senior-zero — COMMENTED** [review]
I've noticed a few changes. I haven't finished the review, will continue once the CI is green.

File: thrust/thrust/detail/tuple.inl [hunk 5]
```diff
@@ -1,1004 +0,0 @@
...
-bool operator<=(const null_type&, const null_type&) { return true; }
-
-__host__ __device__ inline
-bool operator!=(const null_type&, const null_type&) { return false; }
-
-__host__ __device__ inline
-bool operator<(const null_type&, const null_type&) { return false; }
-
-__host__ __device__ inline
-bool operator>(const null_type&, const null_type&) { return false; }
```
Comment (senior-zero) [inline thrust/thrust/detail/tuple.inl:46]: crucial: `thrust::null_type` is defined in public namespace. I don't think we can just remove it. I can see a few projects on github using it.

**Review by senior-zero — CHANGES_REQUESTED** [review]

File: thrust/thrust/tuple.h [hunk 6]
```diff
@@ -178,398 +130,51 @@ get(const detail::cons<HT, TT>& t);
...
-  typename T0, typename T1, typename T2, typename T3, typename T4, typename T5, typename T6, typename T7, typename T8, typename T9,
-  typename U0, typename U1, typename U2, typename U3, typename U4, typename U5, typename U6, typename U7, typename U8, typename U9
->
-inline __host__ __device__
-void swap(tuple<T0,T1,T2,T3,T4,T5,T6,T7,T8,T9> &x,
-          tuple<U0,U1,U2,U3,U4,U5,U6,U7,U8,U9> &y);
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
 
+template<>
+struct tuple_size<tuple<THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<>> {};
```
Comment (griwes) [inline thrust/thrust/tuple.h:154]: For all of these, since we know the number of types we do care about, couldn't we just inherit from integral_constant instead of instantiating tuple_size?

**Review by griwes — APPROVED** [review]
This is great, Thrust's tuple implementation was such an ancient piece of history.

File: thrust/thrust/tuple.h (hunk 6 above)
Comment (miscco) [reply thrust/thrust/tuple.h:154]: This is essentially me being super defensive, as I fear some old compiler will hide a null_type in the initial arguments.

Would that be valid? No
Would it suprise me? No

File: thrust/thrust/detail/tuple.inl (hunk 5 above)
Comment (miscco) [reply thrust/thrust/detail/tuple.inl:46]: Readded null_type and also made sure that `thrust::tuple<T, nulltype>` behaves as expected

File: thrust/thrust/tuple.h [hunk 7]
```diff
@@ -178,398 +130,51 @@ get(const detail::cons<HT, TT>& t);
...
 
+template<class T0>
+struct tuple_size<tuple<T0, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0>> {};
```
Comment (senior-zero) [inline thrust/thrust/tuple.h:157]: suggestion: I can't find tests for tuple_size of tuple<null_type>

**Review by senior-zero — APPROVED** [review]

File: thrust/thrust/tuple.h (hunk 7 above)
Comment (miscco) [reply thrust/thrust/tuple.h:157]: I was pretty sure I added them. will push tomorrow

File: thrust/thrust/tuple.h [hunk 8]
```diff
@@ -125,3 +147,34 @@ using ::cuda::std::tie;
...
+struct tuple_size<tuple<T0, T1, T2, T3, T4, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5, class T6>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, THRUST_NS_QUALIFIER::null_type, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6>> {};
+
+template<class T0, class T1, class T2, class T3, class T4, class T5, class T6, class T7>
+struct tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, T7, THRUST_NS_QUALIFIER::null_type>> : tuple_size<tuple<T0, T1, T2, T3, T4, T5, T6, T7>> {};
```
Comment (runer112) [inline thrust/thrust/tuple.h]: A bit late to the party, but: why does this only go up to 8 non-null elements (0-7)? Shouldn't this go up to 9 (0-8)? AFAICT the old tuple implementation supported up to 10 non-null types, so 9 non-null types plus 1 null type should be handled here?

If true, I think this was unfortunately hidden by the testing added in 471b220d5b46bd4a3163659f532fdeecb6a71d6d falling one element short, as well. There's no test for the case of 9 non-null types followed by `thrust::null_type` having a tuple size of 9. I believe this would currently (incorrectly) produce the standard tuple size of 10.

Comment (runer112) [reply thrust/thrust/tuple.h]: Thinking further, I'm guessing that these specializations were added for compatibility with existing code that fully parameterizes `tuple`? In that case, since they would be specifiying all 10 parameters and all of these specializations only specify 9, none these specializations would match.

When I made my initial comment, I thought that, while only specifying 9 parameters for all of these specializations seemed odd, it was ok because the template parameter defaults specified for `tuple` would provide `thrust::null_type` for the 10th. But that was only true of the old tuple definition, in which case this compatibility layer wouldn't be necessary anyway.

tl;dr: In addition to adding the 9 + 1 specialization, all existing specializations should be extended from 9 to 10 template parameters.

Comment (miscco) [reply thrust/thrust/tuple.h]: 😱 Seems like some of them did not make it
